src/full/Agda/TypeChecking/Irrelevance.hs

```diff
@@ -480,18 +482,53 @@ instance UsableModality a => UsableModality (Arg a) where
 instance UsableModality a => UsableModality (Dom a) where
   usableMod mod Dom{unDom = u} = usableMod mod u
 
-usableAtModality' :: MonadConstraint TCM => Maybe Sort -> Modality -> Term -> TCM ()
-usableAtModality' ms mod t = catchConstraint (UsableAtModality ms mod t) $ do
-  whenM (maybe (pure True) isFibrant ms) $ do
-    res <- runExceptT $ usableMod mod t
-    case res of
-      Right b -> do
-        unless b $
-          typeError . GenericDocError =<< (prettyTCM t <+> "is not usable at the required modality" <+> prettyTCM mod)
-      Left blocker -> patternViolation blocker
-
-
-usableAtModality :: MonadConstraint TCM => Modality -> Term -> TCM ()
+usableAtModality'
+  :: MonadConstraint TCM
+  -- Note: This weird-looking constraint is to trick GHC into accepting
+  -- that an instance of MonadConstraint TCM will exist, even if we
+  -- can't import the module in which it is defined.
+  => Maybe Sort -> WhyCheckModality -> Modality -> Term -> TCM ()
+usableAtModality' ms why mod t =
+  catchConstraint (UsableAtModality why ms mod t) $ do
+    whenM (maybe (pure True) isFibrant ms) $ do
+      res <- runExceptT $ usableMod mod t
+      case res of
+        Right b -> unless b $
+          typeError . GenericDocError =<< formatWhy
+        Left blocker -> patternViolation blocker
+  where
+    formatWhy = do
+      compatible <- collapseDefault . optCubicalCompatible <$> pragmaOptions
+      cubical <- isJust . optCubical <$> pragmaOptions
+      let
+        context
+          | cubical    = "in Cubical Agda,"
+          | compatible = "to maintain compatibility with Cubical Agda,"
+          | otherwise  = "when --without-K is enabled,"
```

Comment:
I don't think I saw all kinds of error messages in the test cases. Please add a test case that uses `--without-K`.